### PR TITLE
Integrate Flux

### PR DESCRIPTION
**Pull request by umiswing**

### CODE REVIEW

File: paddle/phi/infermeta/binary.cc [hunk 1]
```diff
@@ -4197,6 +4197,88 @@ void FakeQuantizeLSQInferMeta(const MetaTensor& x,
...
+          "transpose_weight: %d.",
+          input.dims(),
+          weight.dims(),
+          transpose_weight));
+
+  if (check_can_implement) {
+    output->set_dims(common::make_ddim({}));
+    output->set_dtype(DataType::BOOL);
+  } else {
+    output->set_dtype(input.dtype());
```
Comment (sneaxiy) [inline paddle/phi/infermeta/binary.cc:4278]: Why not invoke `output->set_dims(...)` here?

File: paddle/phi/kernels/gpu/all_gather_gemm_kernel.cu [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+  int32_t nnodes;
+  int32_t full_m;
+  int32_t n_dim;
+  int32_t k_dim;
+  const phi::DataType input_dtype =
+      std::is_same<InT, phi::dtype::float16>::value ? phi::DataType::FLOAT16
+                                                    : phi::DataType::BFLOAT16;
+  const phi::DataType output_dtype =
+      std::is_same<OutT, phi::dtype::float16>::value ? phi::DataType::FLOAT16
+                                                     : phi::DataType::BFLOAT16;
```
Comment (sneaxiy) [inline paddle/phi/kernels/gpu/all_gather_gemm_kernel.cu:59]: 1. Is it supported when `InT` is not float16 or bfloat16?
2. Is it supported when `OutT` is not float16 or bfloat16?
3. Is it supported when `InT` is not the same with `OutT`?

If not, we should add `static_assert` here.

File: paddle/phi/kernels/gpu/all_gather_gemm_kernel.cu [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+  static constexpr int MAX_NUM_SIGNAL = 64;
+
+  using FlagType = int32_t;
+  const phi::GPUContext& dev_ctx;
+  distributed::NCCLCommContext* comm_ctx;
+  paddle::distributed::ProcessGroup* tp_group;
+  int32_t nnodes;
+  int32_t full_m;
+  int32_t n_dim;
+  int32_t k_dim;
```
Comment (sneaxiy) [inline paddle/phi/kernels/gpu/all_gather_gemm_kernel.cu:53]: Note that when `full_m`, `n_dim` or `k_dim` is larger than `2^31-1`, the operator may crash.

File: paddle/phi/kernels/gpu/all_gather_gemm_kernel.cu [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+  AGGemmHelper(const phi::GPUContext& dev_ctx_,
+               paddle::distributed::ProcessGroup* tp_group_,
+               distributed::NCCLCommContext* comm_ctx_,
+               int32_t nnodes,
+               int32_t full_m,
+               int32_t n_dim,
+               int32_t k_dim,
+               bool transpose_weight = true,
+               AGRingMode ring_mode_ = AGRingMode::Auto)
+      : dev_ctx(dev_ctx_),
```
Comment (sneaxiy) [inline paddle/phi/kernels/gpu/all_gather_gemm_kernel.cu:103]: In fact, according to Google coding style, the data member should be named like `dev_ctx_`, and the other member should be named like `dev_ctx`.

File: paddle/phi/kernels/gpu/all_gather_gemm_kernel.cu [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+    this->ring_mode = get_ring_mode(ring_mode_);
+
+    // copy stream
+    this->num_cp_streams = 1;
+    for (int i = 0; i < this->num_cp_streams; ++i) {
+      this->cp_streams.push_back(this->comm_ctx->GetStream());
+    }
+    // create events
+    constexpr bool disable_timing = true;
+    static CUDAEventHolder cp_event_holder{disable_timing};
```
Comment (sneaxiy) [inline paddle/phi/kernels/gpu/all_gather_gemm_kernel.cu:140]: Not a best practice to init the static object in this way. Same below.

File: paddle/phi/kernels/gpu/all_gather_gemm_kernel.cu [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+      }
+    }
+  }
+
+  PADDLE_ENFORCE_GPU_SUCCESS(
+      cudaEventRecord(helper.cp_event, helper.cp_streams[0]));
+
+  ag_gemm();
+
+  PADDLE_ENFORCE(output_buffer.meta().is_contiguous(),
```
Comment (sneaxiy) [inline paddle/phi/kernels/gpu/all_gather_gemm_kernel.cu:376]: `PADDLE_ENFORCE` -> `PADDLE_ENFORCE_EQ(..., true)`

File: paddle/phi/kernels/gpu/all_gather_gemm_kernel.cu (hunk 2 above)
Comment (umiswing) [reply paddle/phi/kernels/gpu/all_gather_gemm_kernel.cu:59]: 1 and 2: No, only bfloat16/float16 and fp8 are supported.

3: There is no constraint in the source code, but it's never tested when `InT` is not the same with `OutT`

File: paddle/phi/infermeta/binary.cc (hunk 1 above)
Comment (umiswing) [reply paddle/phi/infermeta/binary.cc:4278]: Setting output's dims in infermeta require `world_size` passed as parameter. I pass `ring_id` as kernel's parameter and get `world_size` with `ProcessGroup` inside the kernel

**Review by sneaxiy — APPROVED** [review]
Should refine the codes asap.

**Pull request merged**